packages/freezed/test/json_test.dart

```diff
@@ -28,10 +29,38 @@ Future<void> main() async {
       DefaultValue(42),
     );
   });
-  test('@Default does not imply a @JsonKey if one is already specified', () {
+  test('@required implies a @JsonKey', () {
```

Comment:
What is the purpose of this behavior?

Before, it threw a null exception anyway, so I don't see what this adds

Reply:
So before it threw an `AssertionError` -- which is fine in debug builds, but would not get thrown in release builds. This would mean that if deserialization failed in production you would have no way of knowing and the app could end up in a bad state.

What this new behavior does is produce a `CheckedFromJsonException`-- meaning that, since it's not an assertion error, it would still be produced in release builds.

Reply:
That's a fair point, but it's unrelated to deserialisation.
If we want to throw errors on release, this should happen for all usages, not just `.fromJson`

It could be a different mode where instead of:

```dart
MyClass(int foo): assert(foo != null);
```

we get:

```dart
MyClass(int foo) {
  if (foo == null) throw StateError('foo cannot be null');
}
```

Reply:
Yes! I considered that too. It just felt annoying that then we couldn't have `const` constructors anymore 😞 But I hear your point! In the end, and to keep things simpler, I thought that throwing errors in deserialization was more important because those models get initialized from data that comes from the server and we don't control. So the chances of something going wrong there seemed higher than the chances of manually building a model passing in invalid values.

Reply:
If the exception happens during deserialisation, const constructors don't matter

The benefit is, this continues to work on copyWith too

Reply:
Ok, I'm happy to explore that approach instead if you think that's better!